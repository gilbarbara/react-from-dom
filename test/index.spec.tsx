import * as React from 'react';
import { mount } from 'enzyme';

import convert, { convertFromNode, convertFromString } from '../src/index';

import { audio, form, iframe, panel, svg, svgWithStyleAndScript, utf8 } from './__fixtures__/data';

const ReactMarkdown: React.FC = ({ children }) => <div>{children}</div>;

describe('react-from-dom', () => {
  it('should convert an SVG from a string', () => {
    const node = convertFromString(svg, { nodeOnly: true });

    expect(node).toMatchSnapshot();

    // @ts-ignore
    const element: React.ReactNode = convert(node);

    const wrapper = mount(element as React.ReactElement);
    expect(wrapper).toMatchSnapshot();
  });

  it('should convert an SVG with style and script from a string', () => {
    const element: React.ReactNode = convert(svgWithStyleAndScript, {
      selector: 'svg',
    });
    const wrapper = mount(element as React.ReactElement);

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle UTF8 text', () => {
    const element: React.ReactNode = convert(utf8);
    const wrapper = mount(element as React.ReactElement);

    expect(wrapper).toMatchSnapshot();
  });

  it('should convert a search form from a string', () => {
    const element: React.ReactNode = convert(form);
    const wrapper = mount(element as React.ReactElement);

    expect(wrapper).toMatchSnapshot();
  });

  it('should convert an iframe from a string', () => {
    const element: React.ReactNode = convert(iframe);
    const wrapper = mount(element as React.ReactElement);

    expect(wrapper).toMatchSnapshot();
  });

  it('should convert an audio from Node', () => {
    const element: React.ReactNode = convert(audio as Node);
    const wrapper = mount(element as React.ReactElement);

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle actions', () => {
    const element = convert(panel, {
      actions: [
        {
          condition: (node) => node.nodeName.toLowerCase() === 'code',
          // @ts-ignore
          pre: () => null,
        },
        {
          condition: (node) => node.nodeName.toLowerCase() === 'pre',
          post: (node, key) => (
            // @ts-ignore
            <ReactMarkdown key={key} source={node.textContent} />
          ),
        },
        {
          condition: (node) => node.nodeName.toLowerCase() === 'ul',
          pre: (node) => {
            const ol = document.createElement('ol');

            [...node.childNodes].forEach((child) => {
              ol.appendChild(child);
            });

            return ol;
          },
        },
        {
          condition: (node) => node instanceof HTMLElement && node.className === 'panel-footer',
          post: () => null,
        },
        {
          condition: (node) => node instanceof HTMLElement && node.classList.contains('panel'),
          pre: (node) => {
            if (node instanceof HTMLElement) {
              // eslint-disable-next-line no-param-reassign
              node.className += ' panel--fixed';
            }

            return node;
          },
        },
      ],
      selector: 'div',
    });
    const wrapper = mount(element as React.ReactElement);

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle broken markup', () => {
    const element: React.ReactNode = convert('<div><span>los</span>', {
      selector: 'div',
    });
    const wrapper = mount(element as React.ReactElement);

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle missing or invalid parameters', () => {
    // @ts-ignore
    expect(convert()).toBeNull();

    // @ts-ignore
    expect(convert(() => ({}))).toBeNull();

    // @ts-ignore
    expect(convertFromNode()).toBeNull();

    // @ts-ignore
    expect(convertFromNode('This is not a test')).toBeNull();

    // @ts-ignore
    expect(convertFromString()).toBeNull();

    // @ts-ignore
    expect(convertFromString([])).toBeNull();

    // @ts-ignore
    expect(convertFromString('This is a test')).toBeNull();
  });
});
