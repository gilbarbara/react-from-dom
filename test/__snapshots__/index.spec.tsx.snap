// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-from-dom should convert a search form from a string 1`] = `
<form
  _lpchecked="1"
  action="/search"
  autoComplete="off"
  className="search-bar"
  id="search"
  method="get"
>
  <h1
    key="1-3"
  >
    Search
  </h1>
  
  Human
  
  <label
    htmlFor="pet-select"
    key="1-5"
  >
    Choose a pet:
  </label>
  <select
    id="pet-select"
    key="1-7"
  >
    <option
      key="2-1"
      value=""
    >
      --Please choose an option--
    </option>
    <option
      key="2-3"
      value="dog"
    >
      Dog
    </option>
    <option
      key="2-5"
      selected={true}
      value="cat"
    >
      Cat
    </option>
  </select>
  <input
    checked={true}
    key="1-9"
    name="visible"
    type="checkbox"
    value="on"
  />
  <div
    className="ps-relative"
    key="1-11"
  >
    <input
      aria-label="Search"
      autoComplete="off"
      className="s-input s-input__search js-search-field "
      key="2-1"
      maxLength="240"
      name="q"
      placeholder="Search…"
      type="text"
    />
    <svg
      aria-hidden="true"
      className="svg-icon s-input-icon s-input-icon__search iconSearch"
      height="18"
      key="2-3"
      viewBox="0 0 18 18"
      width="18"
    >
      <path
        d="M18 16.5l-5.14-5.18h-.35a7 7 0 1 0-1.19 1.19v.35L16.5 18l1.5-1.5zM12 7A5 5 0 1 1 2 7a5 5 0 0 1 10 0z"
        key="3-0"
      />
    </svg>
  </div>
</form>
`;

exports[`react-from-dom should convert an SVG from a string 1`] = `
<svg
  height="6px"
  preserveAspectRatio="xMidYMid"
  version="1.1"
  viewBox="0 0 20 6"
  width="20px"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  
    
  <defs>
    
        
    <path
      d="M3,6 C1.34314575,6 0,4.65685425 0,3 C0,1.34314575 1.34314575,0 3,0 C4.65685425,0 6,1.34314575 6,3 C6,4.65685425 4.65685425,6 3,6 Z M10,6 C8.34314575,6 7,4.65685425 7,3 C7,1.34314575 8.34314575,0 10,0 C11.6568542,0 13,1.34314575 13,3 C13,4.65685425 11.6568542,6 10,6 Z M17,6 C15.3431458,6 14,4.65685425 14,3 C14,1.34314575 15.3431458,0 17,0 C18.6568542,0 20,1.34314575 20,3 C20,4.65685425 18.6568542,6 17,6 Z"
      id="path-1"
    />
    
        
    <radialgradient
      cx="50%"
      cy="50%"
      fx="50%"
      fy="50%"
      gradientTransform="translate(0.500000,0.500000),scale(0.300000,1.000000),rotate(180.000000),translate(-0.500000,-0.500000)"
      id="radialGradient-3"
      r="166.666667%"
    >
      
            
      <stop
        offset="0%"
        stop-color="#F8E71C"
      />
      
            
      <stop
        offset="100%"
        stop-color="#FF0044"
      />
      
        
    </radialgradient>
    
    
  </defs>
  
    
  <g>
    
        
    <mask
      id="mask-2"
      style="fill: #fff;"
    >
      
            
      <use
        xlink:href="#path-1"
      />
      
        
    </mask>
    
        
    <polygon
      fill="url(#radialGradient-3)"
      fill-rule="nonzero"
      mask="url(#mask-2)"
      points="0 0 20 0 20 6 0 6"
      style="z-index: 100"
    />
    
    
  </g>
  

</svg>
`;

exports[`react-from-dom should convert an SVG from a string 2`] = `
<svg
  height="6px"
  preserveAspectRatio="xMidYMid"
  version="1.1"
  viewBox="0 0 20 6"
  width="20px"
  xmlns="http://www.w3.org/2000/svg"
  xmlnsXlink="http://www.w3.org/1999/xlink"
>
  <defs
    key="1-1"
  >
    <path
      d="M3,6 C1.34314575,6 0,4.65685425 0,3 C0,1.34314575 1.34314575,0 3,0 C4.65685425,0 6,1.34314575 6,3 C6,4.65685425 4.65685425,6 3,6 Z M10,6 C8.34314575,6 7,4.65685425 7,3 C7,1.34314575 8.34314575,0 10,0 C11.6568542,0 13,1.34314575 13,3 C13,4.65685425 11.6568542,6 10,6 Z M17,6 C15.3431458,6 14,4.65685425 14,3 C14,1.34314575 15.3431458,0 17,0 C18.6568542,0 20,1.34314575 20,3 C20,4.65685425 18.6568542,6 17,6 Z"
      id="path-1"
      key="2-1"
    />
    <radialGradient
      cx="50%"
      cy="50%"
      fx="50%"
      fy="50%"
      gradientTransform="translate(0.500000,0.500000),scale(0.300000,1.000000),rotate(180.000000),translate(-0.500000,-0.500000)"
      id="radialGradient-3"
      key="2-3"
      r="166.666667%"
    >
      <stop
        key="3-1"
        offset="0%"
        stopColor="#F8E71C"
      />
      <stop
        key="3-3"
        offset="100%"
        stopColor="#FF0044"
      />
    </radialGradient>
  </defs>
  <g
    key="1-3"
  >
    <mask
      id="mask-2"
      key="2-1"
      style={
        Object {
          "fill": "#fff",
        }
      }
    >
      <use
        key="3-1"
        xlinkHref="#path-1"
      />
    </mask>
    <polygon
      fill="url(#radialGradient-3)"
      fillRule="nonzero"
      key="2-3"
      mask="url(#mask-2)"
      points="0 0 20 0 20 6 0 6"
      style={
        Object {
          "zIndex": 100,
        }
      }
    />
  </g>
</svg>
`;

exports[`react-from-dom should convert an SVG with style and script from a string 1`] = `
<svg
  height="400px"
  preserveAspectRatio="xMidYMid"
  version="1.1"
  viewBox="0 0 600 400"
  width="600px"
  xmlns="http://www.w3.org/2000/svg"
>
  <style
    key="1-1"
  >
    
      @import url(main.css);
    
  </style>
  <style
    key="1-3"
  >
    
      circle:last-child {
        fill: green;
      }
    
  </style>
  <script
    key="1-5"
    type="text/javascript"
  >
    
      
        function change(evt) {
          var target = evt.target;
          var radius = target.getAttribute("r");

          if (radius == 200) {
            radius = 100;
          } else {
            radius = 200;
          }

          target.setAttribute("r", radius);
        }
      
    
  </script>
  <g
    key="1-7"
  >
    <circle
      cx="200"
      cy="200"
      fill="#DDDDDD"
      key="2-1"
      onClick="change(evt)"
      r="200"
    />
    <circle
      cx="400"
      cy="200"
      fill="#000000"
      fillOpacity="0.8"
      key="2-3"
      r="200"
    />
  </g>
</svg>
`;

exports[`react-from-dom should convert an audio from Node 1`] = `
<audio
  controls={true}
  src="https://interactive-examples.mdn.mozilla.net/media/examples/t-rex-roar.mp3"
>
  Your browser does not support the audio element.
</audio>
`;

exports[`react-from-dom should convert an iframe from a string 1`] = `
<iframe
  allow="autoplay; encrypted-media"
  allowFullScreen={true}
  frameBorder="0"
  height="270"
  src="https://www.youtube.com/embed/kemF0xdmsaY?feature=oembed"
  width="480"
/>
`;

exports[`react-from-dom should handle UTF8 text 1`] = `
<svg>
  <tspan
    baselineShift="0%"
    fill="#E4E9F2"
    fontFamily="斯柯达体"
    fontSize="20"
    key="1-1"
    xmlSpace="preserve"
  >
    眼观：仪表盘、车辆
  </tspan>
</svg>
`;

exports[`react-from-dom should handle actions 1`] = `
<div
  className="panel panel--fixed"
>
  <div
    className="panel-header"
    key="1-1"
  >
    <h2
      key="2-1"
    >
      Title
    </h2>
  </div>
  <div
    className="panel-content"
    key="1-3"
  >
    <ol
      key="2-1"
    >
      <li
        key="3-1"
      >
        line 1
      </li>
      <li
        key="3-3"
      >
        line 2
      </li>
    </ol>
    <p
      key="2-3"
    >
      
      This Element has a 
      <code
        key="3-1"
      >
        style
      </code>
       attribute.
      <br
        key="3-3"
      />
      
      It also shows how to use the 
      <code
        key="3-5"
      >
        parser
      </code>
      -argument
    
    </p>
    <p
      key="2-5"
    >
       
    </p>
    <p
      key="2-7"
    >
      
      This is the second paragraph
    
    </p>
    <img
      key="2-9"
    />
    EMPTY
    
    <ReactMarkdown
      key="2-11"
      source="This is a test"
    >
      <div />
    </ReactMarkdown>
    <code
      key="2-13"
    >
      throw new Error('Fail');
    </code>
    <dl
      key="2-15"
    />
  </div>
</div>
`;

exports[`react-from-dom should handle broken markup 1`] = `
<div>
  <span
    key="1-0"
  >
    los
  </span>
</div>
`;
